Source/DafnyLanguageServer/Workspace/Notifications/VerificationDiagnosticsParams.cs

```diff
@@ -0,0 +1,640 @@
+          break;
+        }
+        if (diagnostic.Range.Start.Line >= 0 && diagnostic.Range.Start.Line < result.Length) {
+          result[diagnostic.Range.Start.Line] = LineVerificationStatus.ResolutionError;
+          resolutionErrorRendered++;
+        }
+      }
+
+      var existsErrorRange = false;
+      var existsError = false;
+      foreach (var line in result) {
+        existsErrorRange = existsErrorRange || line == LineVerificationStatus.ErrorContext;
+        existsError = existsError || line == LineVerificationStatus.AssertionFailed;
+      }
+
+      if (existsErrorRange && !existsError) {
+        existsError = false;
+      }
+
+      return result;
+    }
+  }
+
+
+  public enum VerificationStatus {
+    Unknown = 0,
+    Verified = 200,
+    Inconclusive = 270,
+    Error = 400
+  }
+
+  public enum CurrentStatus {
+    Current = 0,
+    Obsolete = 1,
+    Verifying = 2
+  }
+
+  public enum LineVerificationStatus {
+    // Default value for every line, before the renderer figures it out.
+    Unknown = 0,
+    // For first-time computation not actively computing but soon. Synonym of "obsolete"
+    // (scheduledComputation)
+    Scheduled = 1,
+    // For first-time computations, actively computing
+    Verifying = 2,
+    VerifiedObsolete = 201,
+    VerifiedVerifying = 202,
+    // Also applicable for empty spaces if they are not surrounded by errors.
+    Verified = 200,
+    // For trees containing children with errors (e.g. methods)
+    ErrorContextObsolete = 301,
+    ErrorContextVerifying = 302,
+    ErrorContext = 300,
+    // For individual assertions in error ranges
+    AssertionVerifiedInErrorContextObsolete = 351,
+    AssertionVerifiedInErrorContextVerifying = 352,
+    AssertionVerifiedInErrorContext = 350,
+    // For specific lines which have errors on it. They take over verified assertions
+    AssertionFailedObsolete = 401,
+    AssertionFailedVerifying = 402,
+    AssertionFailed = 400,
+    // For lines containing resolution or parse errors
+    ResolutionError = 500
+  }
+
+  public record VerificationTree(
+     // User-facing name
+     string DisplayName,
+     // Used to re-trigger the verification of some diagnostics.
+     string Identifier,
+     string Filename,
+     // The range of this node.
+     Range Range
+  ) {
+    // Overriden by checking children if there are some
+    public VerificationStatus StatusVerification { get; set; } = VerificationStatus.Unknown;
+
+    // Overriden by checking children if there are some
+    public CurrentStatus StatusCurrent { get; set; } = CurrentStatus.Obsolete;
+
+    // Used to relocate a verification tree and to determine which function is currently verifying
+    public Position Position => Range.Start;
+
+    /// Time and Resource diagnostics
+    public bool Started { get; set; } = false;
+    public bool Finished { get; set; } = false;
+    public DateTime StartTime { get; protected set; }
```

Comment:
This information and the other performance related information is currently not used by the client right? I suggest taking it out to reduce the scope of this PR

Reply:
I use `Started` in the `VerificationTree.Start()` method to ensure I set all the children's current status as `Verifying` only once.
I use `Finished` in the `Verification.Stop()` method to ensure I stop only once, because this action replaces Children by the "NewChildren" array.

Other than that, yes, I'll move `ResourceCount`, `StartTime`, `EndTime` and `TimeSpent` and other associated diagnostics metrics to the branch -verification-cover-hover.